packages/contract/src/contract-factory.ts

```
@@ -154,18 +200,135 @@ export default class ContractFactory {
+      const chunk = this.bytecode.slice(offset, offset + chunkSize);
+
+      // Chunks must be byte aligned
+      if (chunk.length % WORD_SIZE !== 0) {
+        const paddingLength = chunk.length % WORD_SIZE;
+        const paddedChunk = concat([chunk, new Uint8Array(paddingLength)]);
+        chunks.push({ id: index, size: paddedChunk.length, bytecode: paddedChunk });
+        // eslint-disable-next-line no-continue
+        continue;
+      }
+      chunks.push({ id: index, size: chunk.length, bytecode: chunk });
+    }
+
+    // Deploy the chunks as blob txs
+    for (const { id, bytecode } of chunks) {
+      const blobRequest = this.blobTransactionRequest({
+        blobBytecode: bytecode,
+        ...deployContractOptions,
+      });
+
+      // Store the blobIds for the loader contract
+      chunks[id].blobId = blobRequest.blobId;
+
+      const fundedBlobRequest = await this.fundTransactionRequest(
+        blobRequest,
+        deployContractOptions
+      );
+
+      let result: TransactionResult<TransactionType.Blob>;
+
+      try {
+        const blobTx = await account.sendTransaction(fundedBlobRequest);
+        result = await blobTx.waitForResult();
+      } catch (err: unknown) {
+        // Core will throw for blobs that have already been uploaded, but the blobId
+        // is still valid so we can use this for the loader contract
+        if ((<FuelError>err).code === ErrorCode.BLOB_ID_ALREADY_UPLOADED) {
+          // TODO: We need to unset the cached utxo as it can be reused
+          // this.account?.provider.cache?.del(UTXO_ID);
+          // eslint-disable-next-line no-continue
+          continue;
+        }
+
+        throw new FuelError(ErrorCode.TRANSACTION_FAILED, 'Failed to deploy contract chunk');
+      }
+
+      if (!result.status || result.status !== TransactionStatus.success) {
+        throw new FuelError(ErrorCode.TRANSACTION_FAILED, 'Failed to deploy contract chunk');
+      }
```

Probably not one for now - but I assume there are some conditions where we'd want to retry the transaction rather than just fail?

---

Going for fail fast at the moment, what scenarios do you think?

---

Correct, the whole deployment will fail. Any blobs already uploaded can be reused without additional cost though. I'll add blob retries to #2917.